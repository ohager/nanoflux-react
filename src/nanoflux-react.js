import React from "react";
import Nanoflux from 'nanoflux';

export function withActions(actions, actionMapper) {
	return function (component) {
		
		return class NanofluxActionsConnector extends React.Component {
		
			static __mapActions() {
				const actionObj = typeof(actions) === 'string' ? Nanoflux.getActions(actions) : actions;
				const actionMap = actionMapper(actionObj);
				let mappedActions = {};
				Object.keys(actionMap).forEach(p => {
					mappedActions[p] = actionMap[p].bind(actionObj);
				});
				
				return mappedActions;
			}
			
			render() {
				let props = Object.assign({}, this.props);
				props.actions = Object.assign({}, props.actions, NanofluxActionsConnector.__mapActions());
				return React.createElement(component, props)
			}
		}
	}
}

export function connect(store, stateToPropsMapper) {
	return function (component) {
		
		return class NanofluxStoreConnector extends React.Component {
			
			static __mapStoreState() {
				let state = {};
				Object.keys(stateToPropsMapper).forEach(p => {
					state[p] = stateToPropsMapper[p]()
				});
				return state;
			}
			
			constructor(props) {
				super(props);
				
				const storeObj = typeof(store) === 'string' ? Nanoflux.getStore(store) : store;
				
				this.state = NanofluxStoreConnector.__mapStoreState();
				this.subscription = storeObj.subscribe(this, this.__onStoreUpdate);
			}
			
			componentWillUnmount() {
				this.subscription.unsubscribe();
			}
			
			__onStoreUpdate() {
				this.setState(NanofluxStoreConnector.__mapStoreState())
			}
			
			
			shouldComponentUpdate(nextState) {
				return nextState !== this.state;
			}
			
			render() {
				return React.createElement(component, Object.assign({}, this.props, this.state))
			}
			
		}
		
	}
}

