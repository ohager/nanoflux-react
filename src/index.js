import {connect, withActions} from './nanoflux-react';

export {withActions, connect};