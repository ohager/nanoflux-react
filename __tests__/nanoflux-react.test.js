import React from 'react';
import {connect} from '../src/index';
import Nanoflux from 'nanoflux';
import {mount, render, shallow} from "enzyme";


// --------------------- Nanoflux Setup -------------------------------

const defaultDispatcher = Nanoflux.getDispatcher();

let testState = {
	test: 'initial',
};

// state selector
const getTestState = () => testState.test;

const storeDescriptor = {
	onTestAction: function (arg) {
		testState.test = arg;
		this.notify(testState);
	},
};

const actionDescriptor = {
	testAction: function (message) {
		this.dispatch('testAction', message);
	},
};

// connect stores to dispatcher
defaultDispatcher.connectTo([
	Nanoflux.createStore('testStore', storeDescriptor),
]);

// setup actions for dispatcher
Nanoflux.createActions('testActions', defaultDispatcher, actionDescriptor);

// --------------------- Nanoflux Setup End ----------------------------


class Test extends React.Component{
	render() {
		return <h2>Test, {`${this.props.testProp}`} </h2>
	}
}

const mapStateToProps = {
	testProp: () => getTestState()
};

describe("nanoflux-react.connect", () => {
		
		it("renders App with mapped state to props using *single* store ", () => {
			
			const testComponent = connect('testStore', mapStateToProps)(Test);
			const wrapper = shallow(React.createElement(testComponent));
			
			expect(wrapper.props().testProp).toBeDefined();
			expect(wrapper.props().testProp).toBe('initial');
			expect(wrapper).toMatchSnapshot()
		})
	}
);


