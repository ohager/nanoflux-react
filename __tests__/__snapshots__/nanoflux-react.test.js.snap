exports[`nanoflux-react.connect renders App with mapped state to props using *single* store  1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function buildPredicate],
    "childrenOfNode": [Function childrenOfNode],
    "findWhereUnwrapped": [Function findWhereUnwrapped]
  },
  "length": 1,
  "node": <function Test() {_classCallCheck(this, Test);return _possibleConstructorReturn(this, (Test.__proto__ || Object.getPrototypeOf(Test)).apply(this, arguments));}
    testProp="initial" />,
  "nodes": Array [
    <function Test() {_classCallCheck(this, Test);return _possibleConstructorReturn(this, (Test.__proto__ || Object.getPrototypeOf(Test)).apply(this, arguments));}
      testProp="initial" />
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <function NanofluxStoreConnector(props) {/* istanbul ignore next */_classCallCheck(this, NanofluxStoreConnector);cov_2g1v1qcf3j.f[9]++;cov_2g1v1qcf3j.s[17]++;var _this2 = _possibleConstructorReturn(this, (NanofluxStoreConnector.__proto__ || Object.getPrototypeOf(NanofluxStoreConnector)).call(this,
					props));

					var storeObj = /* istanbul ignore next */(cov_2g1v1qcf3j.s[18]++, typeof store === 'string' ? /* istanbul ignore next */(cov_2g1v1qcf3j.b[1][0]++, /* istanbul ignore next */_nanoflux2.default.getStore(store)) : /* istanbul ignore next */(cov_2g1v1qcf3j.b[1][1]++, store)); /* istanbul ignore next */cov_2g1v1qcf3j.s[19]++;

					/* istanbul ignore next */_this2.state = NanofluxStoreConnector.__mapStoreState(); /* istanbul ignore next */cov_2g1v1qcf3j.s[20]++;
					/* istanbul ignore next */_this2.subscription = storeObj.subscribe( /* istanbul ignore next */_this2, /* istanbul ignore next */_this2.__onStoreUpdate); /* istanbul ignore next */return _this2;
				} />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NanofluxStoreConnector {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "testProp": "initial"
        },
        "subscription": n {
          "unsubscribe": [Function anonymous]
        },
        "updater": Object {
          "enqueueCallback": [Function enqueueCallback],
          "enqueueCallbackInternal": [Function enqueueCallbackInternal],
          "enqueueElementInternal": [Function enqueueElementInternal],
          "enqueueForceUpdate": [Function enqueueForceUpdate],
          "enqueueReplaceState": [Function enqueueReplaceState],
          "enqueueSetState": [Function enqueueSetState],
          "isMounted": [Function isMounted],
          "validateCallback": [Function validateCallback]
        }
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <function Test() {_classCallCheck(this, Test);return _possibleConstructorReturn(this, (Test.__proto__ || Object.getPrototypeOf(Test)).apply(this, arguments));}
          testProp="initial" />,
        "_debugID": 2,
        "_renderedOutput": <function Test() {_classCallCheck(this, Test);return _possibleConstructorReturn(this, (Test.__proto__ || Object.getPrototypeOf(Test)).apply(this, arguments));}
          testProp="initial" />
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false
    },
    "getRenderOutput": [Function getRenderOutput],
    "render": [Function render]
  },
  "root": [Circular],
  "unrendered": <function NanofluxStoreConnector(props) {/* istanbul ignore next */_classCallCheck(this, NanofluxStoreConnector);cov_2g1v1qcf3j.f[9]++;cov_2g1v1qcf3j.s[17]++;var _this2 = _possibleConstructorReturn(this, (NanofluxStoreConnector.__proto__ || Object.getPrototypeOf(NanofluxStoreConnector)).call(this,
					props));

					var storeObj = /* istanbul ignore next */(cov_2g1v1qcf3j.s[18]++, typeof store === 'string' ? /* istanbul ignore next */(cov_2g1v1qcf3j.b[1][0]++, /* istanbul ignore next */_nanoflux2.default.getStore(store)) : /* istanbul ignore next */(cov_2g1v1qcf3j.b[1][1]++, store)); /* istanbul ignore next */cov_2g1v1qcf3j.s[19]++;

					/* istanbul ignore next */_this2.state = NanofluxStoreConnector.__mapStoreState(); /* istanbul ignore next */cov_2g1v1qcf3j.s[20]++;
					/* istanbul ignore next */_this2.subscription = storeObj.subscribe( /* istanbul ignore next */_this2, /* istanbul ignore next */_this2.__onStoreUpdate); /* istanbul ignore next */return _this2;
				} />
}
`;
